### DISCUSSION TITLE
Cleode ZPLUG

### DISCUSSION THREAD

**Discussion by sivantoledo** [Koenkk/zigbee2mqtt#10693] [Devices]
Hi, I am trying to get an old (~2013) zigbee switched outlet to work. The vendor is Cleode and the model is ZPLUG. Here is the documentation:
http://81.250.221.126:4080/en/produits.php?page=zplug

I got it to join the network and to read attributes using the frontend.
However, it has one endpoint and in zigbee2mqtt it is reported as an input endpoint. I think that it is actually input/output since there is no other endpoint (in the manual). So I can read the on/off attribute but when I try to write the attribute, I get an error:

Publish 'set' 'write' to '0x00124b00012452b0' failed: 'Error: Write 0x00124b00012452b0/1 genOnOff({"onOff":1}, {"sendWhenActive":false,"timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":null,"transactionSequenceNumber":null,"writeUndiv":false}) failed (Status 'READ_ONLY')'

What am I doing wrong? 

Thanks, Sivan Toledo

**Comment by Koenkk** [comment]
The on/off state can be changed by a command, no a write. try to follow https://www.zigbee2mqtt.io/advanced/support-new-devices/01_support_new_devices.html#instructions the following external converter can be used as an example: https://github.com/Koenkk/zigbee2mqtt.io/blob/master/docs/externalConvertersExample/switch.js

**Reply by sivantoledo** [reply]
Thanks a lot. I had a converter almost exactly like the example, but I made a mistake and didn't realize that zigbee2mqtt associates it with the device based on the zigbeeModel property, not on the separate model and manufacturer. Once I changed zigbeeModel to what the log showed, it associated.

Now I can see appropriate log messages when I press the button on the screen, but also an error message that I do not understand:

Zigbee2MQTT:debug 2022-01-10 20:17:13: Received Zigbee message from 'Switched Plug', type 'attributeReport', cluster 'genOnOff', data
'{"onOff":1}' from endpoint 1 with groupID 0
Zigbee2MQTT:info  2022-01-10 20:17:13: Configuring 'Switched Plug'
Zigbee2MQTT:info  2022-01-10 20:17:13: MQTT publish: topic 'zigbee2mqtt/Switched Plug', payload '{"linkquality":116,"state_1":"ON"}'
Zigbee2MQTT:error 2022-01-10 20:17:36: Failed to configure 'Switched Plug', attempt 2 (Error: ConfigureReporting 0x00124b00012452b0/1
genOnOff([{"attribute":"onOff","minimumReportInterval":0,"maximumReportInterval":3600,"reportableChange":0}], {"timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":null,"transactionSequenceNumber":null,"writeUndiv":false}) failed (Timeout - 10421 - 1 - 4 - 6 - 7 after 10000ms)
    at Timeout._onTimeout (C:\Programs\zigbee2mqtt\node_modules\zigbee-herdsman\src\utils\waitress.ts:64:35)
    at listOnTimeout (node:internal/timers:557:17)
    at processTimers (node:internal/timers:500:7))

Also, how do I sent the command? through an mqtt publish event?

**Reply by Koenkk** [reply]
Try changing the configure to:

```js
    configure: async (device, coordinatorEndpoint, logger) => {
        const endpoint = device.getEndpoint(1);
        await reporting.bind(endpoint, coordinatorEndpoint, ['genOnOff']);
    },
```

> Also, how do I sent the command? through an mqtt publish event?

you can check if it works via the z2m frontend -> click on the device -> exposes

**Reply by sivantoledo** [reply]
Works like a charm, both device->exposes and the converter.
I also managed to control it from a command-line mqtt client,
C:\Program Files\mosquitto>mosquitto_pub -t "zigbee2mqtt/Switched Plug/set" -d -m "{"""state""":"""ON"""}"

Very cool.

Many thanks!

**Reply by Koenkk** [reply]
Could you  provide me the external converter so I can add support out-of-the-box?

**Reply by sivantoledo** [reply]
Of course, pasted below. I have a few more Zigbee products by the same manufacturer, I'll try to connect them next week. Also, this one also has an seMetering cluster to report power usage. I'll try to get that to work too. The attributes in that cluster are listed below too.

CurrentSummationDelivered 
Status
UnitofMeasure 
Multiplier 
Divisor
SummationFormatting 
DemandFormatting 
MeteringDeviceType 
InstantaneousDemand
Attribute ID 0x0000
0x0200
0x0300
0x0301
0x0302
0x0303
0x0304
0x0306
0x0400
  
Here is the current converter (on/off only now):

const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const e = exposes.presets;
const ea = exposes.access;

const definition = {
    zigbeeModel: ['ZPLUG'],
    model: 'ZPLUG',
    vendor: 'CLEODE',
    description: 'Cleode ZPLUG mains power switch',
    meta: {multiEndpoint: true},
    fromZigbee: [fz.on_off],
    toZigbee: [tz.on_off],
    exposes: [e.switch()],
    //fromZigbee: [],
    //toZigbee: [],
    //exposes: [],
    // The configure method below is needed to make the device reports on/off state changes
    // when the device is controlled manually through the button on it.
    configure: async (device, coordinatorEndpoint, logger) => {
        const endpoint = device.getEndpoint(1);
        await reporting.bind(endpoint, coordinatorEndpoint, ['genOnOff']);
        // the example had another line but it causes failure to configure the device.
        // await reporting.onOff(endpoint);
    },
};

module.exports = definition;

**Reply by Koenkk** [reply]
try with:

```js
{
        zigbeeModel: ['ZPLUG'],
        model: 'ZPLUG',
        vendor: 'CLEODE',
        description: 'Cleode ZPLUG mains power switch',
        meta: {multiEndpoint: true},
        fromZigbee: [fz.on_off],
        toZigbee: [tz.on_off],
        exposes: [e.switch()],
        configure: async (device, coordinatorEndpoint, logger) => {
            const endpoint = device.getEndpoint(1);
            await reporting.bind(endpoint, coordinatorEndpoint, ['genOnOff', 'seMetering']);
            await reporting.readMeteringMultiplierDivisor(endpoint);
            await reporting.instantaneousDemand(endpoint);
        },
    }
```

**Reply by sivantoledo** [reply]
Yes, this seems to work, but before I got your response I added fz.metering to fromZigbee and e.power() to exploses, and this seemed to work too. I now added the two await statements that you listed and it still seems fine, although I got another configuration error (but everything kept working). The error is

Zigbee2MQTT:error 2022-01-12 20:11:25: Failed to configure 'Switched Plug', attempt 3 (Error: ConfigureReporting 0x00124b00012452b0/1
seMetering([{"attribute":"instantaneousDemand","minimumReportInterval":5,"maximumReportInterval":3600,"reportableChange":1}], {"timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":null,"transactionSequenceNumber":null,"writeUndiv":false}) failed (Timeout - 10421 - 1 - 12 - 1794 - 7 after 10000ms)
    at Timeout._onTimeout (C:\Programs\zigbee2mqtt\node_modules\zigbee-herdsman\src\utils\waitress.ts:64:35)
    at listOnTimeout (node:internal/timers:557:17)
    at processTimers (node:internal/timers:500:7))

and here is the converter I have now:

const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const e = exposes.presets;
const ea = exposes.access;

const definition = {
    zigbeeModel: ['ZPLUG'],
    model: 'ZPLUG',
    vendor: 'CLEODE',
    description: 'Cleode ZPLUG mains power switch',
    meta: {multiEndpoint: true},
    fromZigbee: [fz.on_off,fz.metering],
    toZigbee: [tz.on_off],
    exposes: [e.power(),e.switch()],
    //fromZigbee: [],
    //toZigbee: [],
    //exposes: [],
    // The configure method below is needed to make the device reports on/off state changes
    // when the device is controlled manually through the button on it.
    configure: async (device, coordinatorEndpoint, logger) => {
        const endpoint = device.getEndpoint(1);
        await reporting.bind(endpoint, coordinatorEndpoint, ['genOnOff','seMetering']);
        await reporting.readMeteringMultiplierDivisor(endpoint);
        await reporting.instantaneousDemand(endpoint);
        // the example had another line but it causes failure to configure the device.
        // await reporting.onOff(endpoint);
    },
};

module.exports = definition;

-------------------
The power is now reported in the frontend and it seems very reasonable:

![image](https://user-images.githubusercontent.com/1524989/149198300-220f686e-1321-4cb2-8e3e-94d2d9c4269b.png)

**Reply by Koenkk** [reply]
Should be fixed by removing `await reporting.instantaneousDemand(endpoint);`, added this device now for support out-of-the-box.

Changes will be available in the dev branch in a few hours from now. (https://www.zigbee2mqtt.io/advanced/more/switch-to-dev-branch.html)

**Reply by sivantoledo** [reply]
Koen, there is something weird with the version that I sent: it reports both state and state_1, and I think this prevents the frontend from correctly controlling the switch.

**Reply by Koenkk** [reply]
This is due to `meta: {multiEndpoint: true},` (should be removed) fixed this now.

**Reply by sivantoledo** [reply]
OK, I see that in CLEODE.js. I made the same change in my local copy, I still see both state and state_1 reported, but the frontend work find, publishing a set state. 

I also see weird messaes "No converter available for 'ZPLUG_boost' with cluster genOnOff and type attributeReport and data {onOff: 1} (or data 0).

**Reply by Koenkk** [reply]
> OK, I see that in CLEODE.js. I made the same change in my local copy, I still see both state and state_1 reported, but the frontend work find, publishing a set state.


you can remove state_1 manually from data/state.json

> I also see weird messaes "No converter available for 'ZPLUG_boost' with cluster genOnOff and type attributeReport and data {onOff: 1} (or data 0).

should be fixed now.

Changes will be available in the dev branch in a few hours from now. (https://www.zigbee2mqtt.io/advanced/more/switch-to-dev-branch.html)

**Reply by sivantoledo** [reply]
Yes, fixed, thanks a lot!
